src/main/java/sorald/Repair.java

```diff
@@ -0,0 +1,306 @@
+        intermediateSpoonedPath = spoonedPath.resolve(Constants.INTERMEDIATE);
+    }
+
+    /** Execute a repair according to the config. */
+    public void repair() {
+        List<Integer> ruleKeys = config.getRuleKeys();
+        List<SoraldAbstractProcessor<?>> addedProcessors = new ArrayList<>();
+
+        for (int i = 0; i < ruleKeys.size(); i++) {
+            int ruleKey = ruleKeys.get(i);
+
+            Pair<Path, Path> inOutPaths = computeInOutPaths(i == 0, i == ruleKeys.size() - 1);
+            final Path inputDir = inOutPaths.getLeft();
+            final Path outputDir = inOutPaths.getRight();
+
+            SoraldAbstractProcessor<?> processor = createProcessor(ruleKey);
+            addedProcessors.add(processor);
+            Stream<CtModel> models = repair(inputDir, processor);
+
+            models.forEach(model -> writeModel(model, outputDir));
+        }
+
+        printEndProcess(addedProcessors);
+        FileUtils.deleteDirectory(intermediateSpoonedPath.toFile());
+        UniqueTypesCollector.getInstance().reset();
+    }
+
+    Stream<CtModel> repair(Path inputDir, SoraldAbstractProcessor<?> processor) {
+        if (config.getRepairStrategy() == RepairStrategy.DEFAULT) {
+            CtModel model = defaultRepair(inputDir, processor);
+            return Stream.of(model);
+        } else {
+            assert config.getRepairStrategy() == RepairStrategy.SEGMENT;
+            return segmentRepair(inputDir, processor);
+        }
+    }
+
+    CtModel defaultRepair(Path inputDir, SoraldAbstractProcessor<?> processor) {
+        Launcher launcher = new Launcher();
+        launcher.addInputResource(inputDir.toString());
+        CtModel model = initLauncher(launcher).getModel();
+
+        File inputBaseDir = FileUtils.getClosestDirectory(inputDir.toFile());
+        processor.initResource(inputDir.toString(), inputBaseDir);
+        repairModelWithInitializedProcessor(model, processor);
+
+        return model;
+    }
+
+    Stream<CtModel> segmentRepair(Path inputDir, SoraldAbstractProcessor<?> processor) {
+        Node rootNode = SoraldTreeBuilderAlgorithm.buildTree(inputDir.toString());
+        LinkedList<LinkedList<Node>> segments =
+                FirstFitSegmentationAlgorithm.segment(rootNode, config.getMaxFilesPerSegment());
+        File inputBaseDir = FileUtils.getClosestDirectory(inputDir.toFile());
+
+        return segments.stream()
+                .map(
+                        segment -> {
+                            processor.initResource(segment, inputBaseDir);
+                            Launcher launcher = createSegmentLauncher(segment);
+                            CtModel model = launcher.getModel();
+                            repairModelWithInitializedProcessor(model, processor);
+                            return model;
+                        })
+                .takeWhile(model -> processor.getNbFixes() < config.getMaxFixesPerRule());
+    }
```

Comment:
This is essentially what became of the previous inheritance structure. This is much easier for me to wrap my head around. I envision the public API of Sorald to look something like this.